exchange/bitswap/network/ipfs_impl.go

```diff
@@ -38,15 +40,30 @@ type impl struct {
 }
 
 func (bsnet *impl) newStreamToPeer(ctx context.Context, p peer.ID) (inet.Stream, error) {
-
 	// first, make sure we're connected.
 	// if this fails, we cannot connect to given peer.
 	//TODO(jbenet) move this into host.NewStream?
 	if err := bsnet.host.Connect(ctx, peer.PeerInfo{ID: p}); err != nil {
 		return nil, err
 	}
 
-	return bsnet.host.NewStream(ctx, ProtocolBitswap, p)
+	s, err := bsnet.host.NewStream(ctx, ProtocolBitswap, p)
+	if err != nil {
+		return nil, err
+	}
+
+	// TEMP: trigger read handshake to detect protocl mismatch
+	// TODO: once the network has switched over entirely to the versioned
+	// protocol string, deprecate the old one
+	_, err = s.Read(nil)
+	if err != nil {
+		if strings.Contains(err.Error(), "protocol mismatch") {
+			return bsnet.host.NewStream(ctx, ProtocolBitswapOld, p)
+		}
+		return nil, err
+	}
```

Comment:
i think a better way to do this may be to pay the cost once and then "remember" what this peer has, to prefer that one. that way only the first opening is 1 RTT slow. the way it currently works has 1 RTT cost _per stream opened_.